In [1]:
#Import pandas
import pandas as pd

In [2]:
#Reading a csv file
df = pd.read_csv('bigmart.csv')

In [3]:
#Dimension of the dataframe
df.shape

(8523, 12)

In [4]:
#First 5 rows of the dataframe
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [5]:
#Data types of the variables
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

### One Hot Encoding

In [6]:
df['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

**We can perform One Hot Encoding by using pd.get_dummies() and OneHotEncoder().**

In [7]:
#Perform One Hot Encoding using pd.get_dummies()
df_encoded = pd.get_dummies(df)
df_encoded.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
0         9.30         0.016047  249.8092                       1999   
1         5.92         0.019278   48.2692                       2009   
2        17.50         0.016760  141.6180                       1999   
3        19.20         0.000000  182.0950                       1998   
4         8.93         0.000000   53.8614                       1987   

   Item_Outlet_Sales  Item_Identifier_DRA12  Item_Identifier_DRA24  \
0          3735.1380                      0                      0   
1           443.4228                      0                      0   
2          2097.2700                      0                      0   
3           732.3800                      0                      0   
4           994.7052                      0                      0   

   Item_Identifier_DRA59  Item_Identifier_DRB01  Item_Identifier_DRB13  ...  \
0                      0                      0                      0  ...   
1                      0                      0                      0  ...   
2                      0                      0                      0  ...   
3                      0                      0                      0  ...   
4                      0                      0                      0  ...   

   Outlet_Size_High  Outlet_Size_Medium  Outlet_Size_Small  \
0                 0                   1                  0   
1                 0                   1                  0   
2                 0                   1                  0   
3                 0                   0                  0   
4                 1                   0                  0   

   Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
0                            1                            0   
1                            0                            0   
2                            1                            0   
3                            0                            0   
4                            0                            0   

   Outlet_Location_Type_Tier 3  Outlet_Type_Grocery Store  \
0                            0                          0   
1                            1                          0   
2                            0                          0   
3                            1                          1   
4                            1                          0   

   Outlet_Type_Supermarket Type1  Outlet_Type_Supermarket Type2  \
0                              1                              0   
1                              0                              1   
2                              1                              0   
3                              0                              0   
4                              1                              0   

   Outlet_Type_Supermarket Type3  
0                              0  
1                              0  
2                              0  
3                              0  
4                              0  

[5 rows x 1605 columns]

In [8]:
from sklearn.preprocessing import OneHotEncoder

In [9]:
encoder = OneHotEncoder()
df_encoded = encoder.fit_transform(df[['Outlet_Type']])
df_encoded = pd.DataFrame(df_encoded.toarray())
df_encoded

0    1    2    3
0     0.0  1.0  0.0  0.0
1     0.0  0.0  1.0  0.0
2     0.0  1.0  0.0  0.0
3     1.0  0.0  0.0  0.0
4     0.0  1.0  0.0  0.0
...   ...  ...  ...  ...
8518  0.0  1.0  0.0  0.0
8519  0.0  1.0  0.0  0.0
8520  0.0  1.0  0.0  0.0
8521  0.0  0.0  1.0  0.0
8522  0.0  1.0  0.0  0.0

[8523 rows x 4 columns]

In [10]:
df['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

### Doing One Hot Encoding for all the variables

In [11]:
df_encoded = pd.get_dummies(df)
df_encoded.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
0         9.30         0.016047  249.8092                       1999   
1         5.92         0.019278   48.2692                       2009   
2        17.50         0.016760  141.6180                       1999   
3        19.20         0.000000  182.0950                       1998   
4         8.93         0.000000   53.8614                       1987   

   Item_Outlet_Sales  Item_Identifier_DRA12  Item_Identifier_DRA24  \
0          3735.1380                      0                      0   
1           443.4228                      0                      0   
2          2097.2700                      0                      0   
3           732.3800                      0                      0   
4           994.7052                      0                      0   

   Item_Identifier_DRA59  Item_Identifier_DRB01  Item_Identifier_DRB13  ...  \
0                      0                      0                      0  ...   
1                      0                      0                      0  ...   
2                      0                      0                      0  ...   
3                      0                      0                      0  ...   
4                      0                      0                      0  ...   

   Outlet_Size_High  Outlet_Size_Medium  Outlet_Size_Small  \
0                 0                   1                  0   
1                 0                   1                  0   
2                 0                   1                  0   
3                 0                   0                  0   
4                 1                   0                  0   

   Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
0                            1                            0   
1                            0                            0   
2                            1                            0   
3                            0                            0   
4                            0                            0   

   Outlet_Location_Type_Tier 3  Outlet_Type_Grocery Store  \
0                            0                          0   
1                            1                          0   
2                            0                          0   
3                            1                          1   
4                            1                          0   

   Outlet_Type_Supermarket Type1  Outlet_Type_Supermarket Type2  \
0                              1                              0   
1                              0                              1   
2                              1                              0   
3                              0                              0   
4                              1                              0   

   Outlet_Type_Supermarket Type3  
0                              0  
1                              0  
2                              0  
3                              0  
4                              0  

[5 rows x 1605 columns]

We have two problems here: 

#### Problem 1:
Look at the newly created variables ``Outlet_Size_High``, 	``Outlet_Size_Medium``, 	``Outlet_Size_Small``, the order between these variables is destroyed. As a result we are missing out on some important information.  

In [12]:
df_encoded[['Outlet_Size_High', 'Outlet_Size_Medium','Outlet_Size_Small']].head()

Outlet_Size_High  Outlet_Size_Medium  Outlet_Size_Small
0                 0                   1                  0
1                 0                   1                  0
2                 0                   1                  0
3                 0                   0                  0
4                 1                   0                  0

#### Problem 2
The number of features has increased from 12 to 1605, where maximum values are 0.

In [13]:
df.shape, df_encoded.shape

((8523, 12), (8523, 1605))

In [14]:
df_encoded[['Item_Identifier_DRA12', 'Item_Identifier_DRA24',
       'Item_Identifier_DRA59', 'Item_Identifier_DRB01',
       'Item_Identifier_DRB13', 'Item_Identifier_DRB24',
       'Item_Identifier_DRB25', 'Item_Identifier_DRB48',
       'Item_Identifier_DRC01', 'Item_Identifier_DRC12',
       'Item_Identifier_DRC13', 'Item_Identifier_DRC24',
       'Item_Identifier_DRC25', 'Item_Identifier_DRC27',
       'Item_Identifier_DRC36', 'Item_Identifier_DRC49',
       'Item_Identifier_DRD01', 'Item_Identifier_DRD12',
       'Item_Identifier_DRD13', 'Item_Identifier_DRD15',
       'Item_Identifier_DRD24', 'Item_Identifier_DRD25',
       'Item_Identifier_DRD27', 'Item_Identifier_DRD37',
       'Item_Identifier_DRD49']].head()

Item_Identifier_DRA12  Item_Identifier_DRA24  Item_Identifier_DRA59  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRB01  Item_Identifier_DRB13  Item_Identifier_DRB24  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRB25  Item_Identifier_DRB48  Item_Identifier_DRC01  \
0                      0                      0                      0   
1                      0                      0                      1   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRC12  ...  Item_Identifier_DRC49  Item_Identifier_DRD01  \
0                      0  ...                      0                      0   
1                      0  ...                      0                      0   
2                      0  ...                      0                      0   
3                      0  ...                      0                      0   
4                      0  ...                      0                      0   

   Item_Identifier_DRD12  Item_Identifier_DRD13  Item_Identifier_DRD15  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRD24  Item_Identifier_DRD25  Item_Identifier_DRD27  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRD37  Item_Identifier_DRD49  
0                      0                      0  
1                      0                      0  
2                      0                      0  
3                      0                      0  
4                      0                      0  

[5 rows x 25 columns]

### Solution of Problem 1 : Label Encoding

In [15]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [16]:
df['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [17]:
#Apply LabelEncoder
le.fit_transform(['Small', 'Medium', 'High'])

array([2, 1, 0], dtype=int64)

**LabelEncoder uses the alphabetical order**

In [18]:
df['Outlet_Size'] = df['Outlet_Size'].map({'Small' : 0, 'Medium' : 1, 'High' : 2})

In [19]:
df['Outlet_Size'].head()

0    1.0
1    1.0
2    1.0
3    NaN
4    2.0
Name: Outlet_Size, dtype: float64

So this solves the first challenge we encountered. Now we'll see how to deal with high cardinality.

### Solution of Problem 2

In [20]:
df.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

In [21]:
df['Item_Identifier'].value_counts()

FDW13    10
FDG33    10
FDU12     9
FDT07     9
FDG09     9
         ..
FDY43     1
FDC23     1
DRF48     1
FDK57     1
FDN52     1
Name: Item_Identifier, Length: 1559, dtype: int64

In [22]:
temp = df['Item_Identifier'].value_counts()
temp.head()

FDW13    10
FDG33    10
FDU12     9
FDT07     9
FDG09     9
Name: Item_Identifier, dtype: int64

In [23]:
#Calculate a frequency of each item and store it into a new column
df['Item_Identifier_Count'] = df['Item_Identifier'].apply(lambda x : temp[x])
df[['Item_Identifier', 'Item_Identifier_Count']].head()

Item_Identifier  Item_Identifier_Count
0           FDA15                      8
1           DRC01                      6
2           FDN15                      7
3           FDX07                      6
4           NCD19                      6

In [24]:
for i in range(0, len(df)):
    if df['Item_Identifier_Count'][i] < 4:
        df['Item_Identifier'][i] = 'Other'

<ipython-input-24-2ce77fc59495>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Item_Identifier'][i] = 'Other'


In [25]:
df.head(10)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15        9.300          Low Fat         0.016047   
1           DRC01        5.920          Regular         0.019278   
2           FDN15       17.500          Low Fat         0.016760   
3           FDX07       19.200          Regular         0.000000   
4           NCD19        8.930          Low Fat         0.000000   
5           Other       10.395          Regular         0.000000   
6           FDO10       13.650          Regular         0.012741   
7           FDP10          NaN          Low Fat         0.127470   
8           FDH17       16.200          Regular         0.016687   
9           FDU28       19.200          Regular         0.094450   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   
5           Baking Goods   51.4008            OUT018   
6            Snack Foods   57.6588            OUT013   
7            Snack Foods  107.7622            OUT027   
8           Frozen Foods   96.9726            OUT045   
9           Frozen Foods  187.8214            OUT017   

   Outlet_Establishment_Year  Outlet_Size Outlet_Location_Type  \
0                       1999          1.0               Tier 1   
1                       2009          1.0               Tier 3   
2                       1999          1.0               Tier 1   
3                       1998          NaN               Tier 3   
4                       1987          2.0               Tier 3   
5                       2009          1.0               Tier 3   
6                       1987          2.0               Tier 3   
7                       1985          1.0               Tier 3   
8                       2002          NaN               Tier 2   
9                       2007          NaN               Tier 2   

         Outlet_Type  Item_Outlet_Sales  Item_Identifier_Count  
0  Supermarket Type1          3735.1380                      8  
1  Supermarket Type2           443.4228                      6  
2  Supermarket Type1          2097.2700                      7  
3      Grocery Store           732.3800                      6  
4  Supermarket Type1           994.7052                      6  
5  Supermarket Type2           556.6088                      3  
6  Supermarket Type1           343.5528                      8  
7  Supermarket Type3          4022.7636                      7  
8  Supermarket Type1          1076.5986                      7  
9  Supermarket Type1          4710.5350                      5

In [26]:
df['Item_Identifier'].value_counts()

Other    418
FDW13     10
FDG33     10
NCJ30      9
FDX04      9
        ... 
FDI58      4
NCJ06      4
NCN17      4
FDU59      4
FDV35      4
Name: Item_Identifier, Length: 1403, dtype: int64